### PR TITLE
feat: refreshable query atom

### PR DESCRIPTION
**Pull request by tien**
Small change to allow refreshing of query atom.

### CODE REVIEW

**Comment by dai-shi** [comment]
Hi, thanks for the suggestion. I have a few questions:
1. Does it really refresh (refetch) with this change?
2. Is it a typical way for tRPC for refetching? What is the case with trpc/react binding?

**Comment by tien** [comment]
1. Yes, I tested this quickly with the Pokemon example app. The atom can be refreshed as expected
2. A refresh function is needed if the query result from TRPC can change overtime, for example if the result change after a mutation.

**Comment by dai-shi** [comment]
2: My question is how the API should be designed. I'm afraid the atomWithRefresh usage is not explicit, so people wouldn't notice. Maybe it's just a doc issue.

**Comment by dai-shi** [comment]
https://trpc.io/
Do you know where does it describe about refresh capability in their docs?

**Comment by tien** [comment]
> 2: My question is how the API should be designed. I'm afraid the atomWithRefresh usage is not explicit, so people wouldn't notice. Maybe it's just a doc issue.

Yeah, up to you how you want to expose this to the user, I guess :)

> https://trpc.io/ Do you know where does it describe about refresh capability in their docs?

Oh, refresh here basically mean just fetching the resource again, so for example:

```ts
// Assume the TRPC server expose a query called
// `getCurrentTime` that will just return the current time
const [time, refreshTime] = jotaiClient.getCurrentTime();

// This would change everytime refreshTime is called
console.log(time);
```

**Comment by dai-shi** [comment]
How would you refetch query, if you used `createTRPCReact` without Jotai?

**Comment by tien** [comment]
> How would you refetch query, if you used `createTRPCReact` without Jotai?

`createTRPCReact` match `reactQuery` behaviour, so you will get a `refetch` function that you can call: https://tanstack.com/query/latest/docs/framework/react/reference/useQuery

**Comment by dai-shi** [comment]
> https://tanstack.com/query/latest/docs/framework/react/reference/useQuery

I see.

> Yeah, up to you how you want to expose this to the user, I guess :)

Doesn't sound too bad then. After all, it's still pre v1.

File: src/createTRPCJotai.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ import type { inferObservableValue } from '@trpc/server/observable';
 
 import { atom } from 'jotai/vanilla';
 import type { Atom, Getter, WritableAtom } from 'jotai/vanilla';
-import { atomWithObservable } from 'jotai/vanilla/utils';
+import { atomWithObservable, atomWithRefresh } from 'jotai/vanilla/utils';
```
Comment (dai-shi) [inline src/createTRPCJotai.ts:18]: > "jotai": ">=1.11.0"

`atomWithRefresh` isn't available in older versions. Maybe, we should implement it without it.

File: src/createTRPCJotai.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ const atomWithQuery = <TProcedure extends AnyQueryProcedure, TClient>(
   getOptions?: ValueOrGetter<TRPCRequestOptions & CustomOptions>,
 ) => {
   type Output = inferProcedureOutput<TProcedure>;
-  const queryAtom = atom(async (get, { signal }) => {
+  const queryAtom = atomWithRefresh(async (get, { signal }) => {
```
Comment (dai-shi) [inline src/createTRPCJotai.ts:50]:
```suggestion
  const refreshAtom = atom(0)
  const queryAtom = atom(async (get, { signal }) => {
    get(refreshAtom)
```

**Review by dai-shi — CHANGES_REQUESTED** [review]
Can you refactor it please?

**Comment by dai-shi** [comment]
Please also change this:
```diff
diff --git a/.codesandbox/ci.json b/.codesandbox/ci.json
index a59d926..6afcb06 100644
--- a/.codesandbox/ci.json
+++ b/.codesandbox/ci.json
@@ -1,5 +1,5 @@
 {
   "buildCommand": "compile",
   "sandboxes": ["new", "react-typescript-react-ts"],
-  "node": "14"
+  "node": "18"
 }
```

File: src/createTRPCJotai.ts (hunk 1 above)
Comment (tien) [reply src/createTRPCJotai.ts]: Okay, I've implemented the refresh feature manually like you suggested

**Review by dai-shi — APPROVED** [review]
The code looks good to me!

Can you open another PR in `jotai` repo to update the docs? https://jotai.org/docs/extensions/trpc

**Pull request merged**